packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -27,138 +28,433 @@ import { parseVatSlot } from './parseVatSlots.js';
-      // time we get here. The first step is to filter this down to a
-      // list of definitely dead baserefs.
+      // process a sorted list of vref/baseRefs we need to check for
+      // reachability, one at a time
 
-      const deadSet = new Set();
+      for (const vrefOrBaseRef of [...possiblyDeadSet].sort()) {
+        // remove the vref now, but some deleteVirtualObject might
+        // shake it loose again for a future pass to investigate
+        possiblyDeadSet.delete(vrefOrBaseRef);
+
+        const parsed = parseVatSlot(vrefOrBaseRef);
+        assert.equal(parsed.type, 'object', vrefOrBaseRef);
+
+        let res = {};
+        if (parsed.virtual || parsed.durable) {
+          const baseRef = vrefOrBaseRef;
+          res = checkExportRepresentative(baseRef);
+        } else if (parsed.allocatedByVat) {
+          const vref = vrefOrBaseRef;
+          res = checkExportRemotable(vref);
+        } else {
+          const vref = vrefOrBaseRef;
+          res = checkImportPresence(vref);
+        }
 
-      for (const baseRef of possiblyDeadSet) {
-        if (slotToVal.has(baseRef)) {
-          continue; // RAM pillar remains
+        // prepare our end-of-crank syscalls
+        if (res.dropImport) {
+          importsToDrop.add(res.dropImport);
+          possiblyRetiredSet.add(res.dropImport);
         }
-        const { virtual, durable, type } = parseVatSlot(baseRef);
-        assert(type === 'object', `unprepared to track ${type}`);
-        if (virtual || durable) {
-          if (vrm.isVirtualObjectReachable(baseRef)) {
-            continue; // vdata or export pillar remains
-          }
+        for (const facetVref of res.exportsToRetire || []) {
+          exportsToRetire.add(facetVref);
         }
-        deadSet.add(baseRef);
+        gcAgain ||= !!res.gcAgain;
       }
-      possiblyDeadSet.clear();
-
-      // deadSet now contains objects which are certainly dead
 
-      // possiblyRetiredSet holds (a subset of??) baserefs which have
-      // lost a recognizer recently. TODO recheck this
+      // Deleting virtual object state, or freeing weak-keyed
+      // collection entries, might shake loose more
+      // objects. possiblyDeadSet and possiblyRetiredSet are added
+      // when vdata vrefs are decreffed to zero, and gcAgain means that
+      // something in RAM might now be free. In both cases we should
+      // do another pass, including gcAndFinalize(), until we've
+      // cleared everything we can.
+    } while (possiblyDeadSet.size > 0 || gcAgain);
 
-      for (const vref of possiblyRetiredSet) {
-        if (!getValForSlot(vref) && !deadSet.has(vref)) {
-          // Don't retire things that haven't yet made the transition to dead,
-          // i.e., always drop before retiring
+    // Now we process potential retirements. addToPossiblyRetiredSet()
+    // is given vrefs (of all sorts) for the keys of deleted weak
+    // collections, plus any o-NN "Presence-type" vrefs that were
```

Comment:
My reading is that `addToPossiblyRetiredSet()` is only called for `vref` which are `!allocatedByVat`, so the set should only contain imported presences.

Btw, it's implied here, but it'd be great if the comment could make clear that `addToPossiblyRetiredSet()` might have been called before/outside of this `scanForDeadObject` logic, e.g. in previous deliveries from the kernel.

Reply:
Yeah I decided to have `scanForDeadObjects()` be responsible for filtering the list down to Presence-style (`!allocatedByVat`) vrefs, and at the time I wrote this loop, the set did indeed contain vrefs of all sorts. In the subsequent PR, while I was changing (at least some of) the callers of `addToPossiblyRetiredSet`, I figured it was a perf improvement if they could also do their own filtering. So I'll call it a belt-and-suspenders things.

I'll add that comment, thanks.